R/colour_nhsbsa.R

```diff
@@ -14,54 +15,45 @@
 #' nhsbsaR::palette_nhsbsa(palette = "gender")
 #' nhsbsaR::palette_nhsbsa(palette = "gradient")
 #' nhsbsaR::palette_nhsbsa(palette = "highlight")
+#' #' nhsbsaR::palette_nhsbsa(palette = NHSRtheme::get_nhs_colours(c("Blue", "AquaGreen")))
 #' @export
 palette_nhsbsa <- function(palette = NA, reverse = FALSE) {
-
-  # Depending on the palette, get the colour names
-  if (!is.na(palette)) {
+  # Check if palette is one of the named palettes or a custom color palette
+  if (is.character(palette) & length(palette) == 1) { # Ensure palette is a single string
     if (palette == "gender") {
       names <- c("Pink", "LightBlue")
-    }
-
-    if (palette == "gradient") {
-      names <- c("White", "DarkBlue") # White is custom
-    }
-
-    if (palette == "highlight") {
+    } else if (palette == "gradient") {
+      names <- c("White", "DarkBlue")
+    } else if (palette == "highlight") {
       names <- c("MidGrey", "DarkBlue")
     }
-  } else {
-
-    # Base on the Wong palette (https://davidmathlogic.com/colorblind/)
+  } else if (is.character(palette) & length(palette) > 1) {
+    # If palette is not one of the named palettes, treat as custom colours
+    return(unname(palette))
+  } else if (is.na(palette)) {
+    # Base on the Wong palette
     names <- c(
-      "DarkBlue", # Instead of black
-      "Orange",
-      "LightBlue",
-      "AquaGreen",
-      "Yellow",
-      "BrightBlue", # Instead of NHS Blue
-      "Red",
-      "Pink"
+      "DarkBlue", "Orange", "LightBlue", "AquaGreen",
+      "Yellow", "BrightBlue", "Red", "Pink"
     )
+  } else {
+    stop("Invalid palette specified.")
   }
 
   # Reverse names if necessary
   if (reverse) {
     names <- rev(names)
   }
 
-  # Get the NHS identity hex colour for each name (with a couple of custom
-  # colours added)
-  colours <- c(
-    NHSRtheme::get_nhs_colours(),
-    "White" = "#FFFFFF"
-  )[names]
+  # Get the NHS identity hex colour for each name (with custom colours added)
+  colours <- c(NHSRtheme::get_nhs_colours(), "White" = "#FFFFFF")[names]
 
   # Return the unnamed colours
   unname(colours)
 }
 
 
+
```

Comment:
I have added functionality and checks to allow non-NHS identity colours to also be used.
Renamed the `names` variable to `color_names` , since `names` is an R function (esp. since I have used it too!).
Used a `switch` statement for the known words, more concise and easier to read (IMO).
I think I tried all the cases, but could be some I missed in code below functions.

```
is_color <- function(x) {
  x %in% colors() |
    grepl(
      "^#(\\d|[a-f]){6,8}$",
      x,
      ignore.case=TRUE
    )
}

nhs_colours <- c(NHSRtheme::get_nhs_colours(), "White" = "#FFFFFF")

is_nhs_color <- function(x) {
  all(x %in% names(nhs_colours))
}

palette_nhsbsa <- function(palette = NA, reverse = FALSE) {
  # Check if palette is one of the named palettes or a custom color palette
  if (is.character(palette) & length(palette) == 1) { # Ensure palette is a single string
    color_names <- switch (palette,
      gender = c("Pink", "LightBlue"),
      gradient = c("White", "DarkBlue"),
      highlight = c("MidGrey", "DarkBlue"),
      ifelse(
        is_color(palette) | is_nhs_color(palette),
        return(unname(palette)),
        stop("Invalid palette specified.")
      )
    )
  } else if (is.character(palette) & length(palette) > 1) {
    # If palette is not one of the named palettes, treat as custom colours
    ifelse(
      all(is_color(palette) | all(is_nhs_color(palette))),
      return(unname(palette)),
      stop("Invalid palette specified.")
    )
  } else if (is.na(palette)) {
    # Base on the Wong palette
    color_names <- c(
      "DarkBlue", "Orange", "LightBlue", "AquaGreen",
      "Yellow", "BrightBlue", "Red", "Pink"
    )
  } else {
    stop("Invalid palette specified.")
  }
  
  # Reverse names if necessary
  if (reverse) {
    color_names <- rev(color_names)
  }
  
  if (is_nhs_color(color_names)) {
    unname(nhs_colours[color_names])
  } else {
    unname(nhs_colours)
  }
}

# Works
palette_nhsbsa("gender") # known word
palette_nhsbsa("green") # single R color
palette_nhsbsa(c("blue", "red", "#bada55")) # mixture of R colors and hexes
palette_nhsbsa(c("Orange", "Pink")) # all NHS colour names
palette_nhsbsa("#BADA55") # single hex
palette_nhsbsa("#BADA55CC", "#BADDAD") # multiple hexes, one with alpha characters

# Does not work
palette_nhsbsa("gander") # unknown word
palette_nhsbsa(c("indigo", "Pink", "DarkBlue")) # mixtures not allowed with NHS colours
```